src/main/java/org/opensearch/flowframework/workflow/AbstractRetryableWorkflowStep.java

```diff
@@ -65,82 +70,92 @@ protected AbstractRetryableWorkflowStep(
+                    switch (response.getState()) {
+                        case COMPLETED:
+                            try {
+                                String resourceName = getResourceByWorkflowStep(getName());
+                                String id = getResourceId(response);
+                                logger.info("{} successful for {} and {} {}", workflowStep, workflowId, resourceName, id);
+                                flowFrameworkIndicesHandler.updateResourceInStateIndex(
+                                    workflowId,
+                                    nodeId,
+                                    getName(),
+                                    id,
+                                    ActionListener.wrap(updateResponse -> {
+                                        logger.info("successfully updated resources created in state index: {}", updateResponse.getIndex());
+                                        future.complete(
+                                            new WorkflowData(
+                                                Map.ofEntries(
+                                                    Map.entry(resourceName, id),
+                                                    Map.entry(REGISTER_MODEL_STATUS, response.getState().name())
+                                                ),
+                                                workflowId,
+                                                nodeId
+                                            )
+                                        );
+                                    }, exception -> {
+                                        logger.error("Failed to update new created resource", exception);
+                                        future.completeExceptionally(
+                                            new FlowFrameworkException(exception.getMessage(), ExceptionsHelper.status(exception))
+                                        );
+                                    })
+                                );
+                            } catch (Exception e) {
+                                logger.error("Failed to parse and update new created resource", e);
+                                future.completeExceptionally(new FlowFrameworkException(e.getMessage(), ExceptionsHelper.status(e)));
+                            }
+                            break;
+                        case FAILED:
+                        case COMPLETED_WITH_ERROR:
+                            String errorMessage = workflowStep + " failed with error : " + response.getError();
+                            logger.error(errorMessage);
+                            future.completeExceptionally(new FlowFrameworkException(errorMessage, RestStatus.BAD_REQUEST));
+                            break;
+                        case CANCELLED:
+                            logger.error("{} task was cancelled.", workflowStep);
+                            FutureUtils.cancel(future);
```

Comment:
unlike other cases we don't `completeExceptionally` here, does this mean we don't propagate an exception to update the state index with? Not sure if you tested this, just cause we are treating `cancelled` differently here

Reply:
Cancelling a future will complete it, which either `join()` or `get()` is waiting on.   Since it is not completed normally, it will throw a `CancellationException` (unchecked) on either of those methods.  

The future completed here is the one returned from the associated WorkflowStep subclass to the ProcessNode where it is then returned to the `ProvisionWorkflowTransportAction` (and corresponding deprovision equivalent).  That calls `join` on all the step futures, so will throw an exception on this line.
```java
workflowFutureList.forEach(CompletableFuture::join);
```

This will be caught in the same catch block as being completed exceptionally, but with the different exception type.  

Since the `CancellationException` thrown from the cancelled future won't have a message, we may want to actually test the exception type in that catch block and give a different response.  WDYT?

Reply:
Handled the cancellation (and throwable with cause) error messages in https://github.com/opensearch-project/flow-framework/pull/298/commits/13426e4947e70f71ffb50ec1786d4657a6055e5c

Reply:
Yes this looks good to me so we can make sure that we update the error message with either the error or in case of cancellation some sort of indication that a cancellation happened, I wonder if we can improve though by propagating what step was cancelled, for now I am okay with it being generic like in your commit.

Reply:
Just completed exceptionally instead of cancelling in https://github.com/opensearch-project/flow-framework/pull/298/commits/a0fc9834f16685b064530d5159ee59df65a33efd